### PR TITLE
[FIX] sale: fix wrong amount with downpayment

### PR DESCRIPTION
**Pull request by guva-odoo**
Steps to reproduce:

- Create and confirm a sale order for a product with for example
  a price of 250$ with 15% tax (Total = 287.50$)
- Create and confirm a downpayment for 100$, with 15% tax (Total = 115$)
- Create an invoice for the rest of the outstanding amount,with downpayment deducted.
- Confirm, then reset to draft and cancel entry
- Create a new invoice for the outstanding amount

Issue:

The last invoice's total amount is equal to the order's total amount,
which is not correct, it should be order's total amount - downpayment
(284.50 - 115.00 = 172.50$)

The reason is when confirm the second invoice, the price od dosnpayment line
is computed to zero, but never reset to initial value when reseting the invoice to draft.

Solution:

Overwritte of the button_draft method in order to reverse what
action_post() did.

opw-2694101

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/sale/models/account_move.py [hunk 1]
```diff
@@ -28,6 +28,21 @@ def action_post(self):
                 pass
         return res
 
+    def button_draft(self):
+        #inherit of the function from account.move to validate a new tax and the priceunit of a downpayment
```
Comment (agr-odoo) [inline addons/sale/models/account_move.py:32]: maybe change this to suit more the use case

File: addons/sale/models/account_move.py [hunk 2]
```diff
@@ -28,6 +28,21 @@ def action_post(self):
...
+        res = super(AccountMove, self).button_draft()
```
Comment (agr-odoo) [inline addons/sale/models/account_move.py:33]: `button_draft` does not return, so it's pointless to save the return value

File: addons/sale/models/account_move.py [hunk 3]
```diff
@@ -28,6 +28,21 @@ def action_post(self):
...
+        line_ids = self.mapped('line_ids').filtered(lambda line: line.sale_line_ids.is_downpayment)
+        for line in line_ids:
+            try:
+                line.sale_line_ids.tax_id = line.tax_ids
```
Comment (agr-odoo) [inline addons/sale/models/account_move.py:37]: this was already set on the `action_post`, it seems to be untouched in the mentioned flow so it's not needed the reassignment

**Review by agr-odoo — APPROVED** [review]
Looks ok, beside the small changes I noted
@Feyensv Does this looks ok to you?

**Comment by Feyensv** [comment]
Honestly, it's quite an obscure hack, I wouldn't dare to confirm/refuse an accounting change like this one :thinking:.
Do you mind confirming with the accounting team? (since it's more in the accounting part of sale).

**Comment by agr-odoo** [comment]
Hello @Levizar 
May we have an accountant opinion for this?

File: addons/sale/models/account_move.py [hunk 4]
```diff
@@ -28,6 +28,19 @@ def action_post(self):
...
+        super(AccountMove, self).button_draft()
+        line_ids = self.mapped('line_ids').filtered(lambda line: line.sale_line_ids.is_downpayment)
+        for line in line_ids:
+            try:
+                if not all(line.tax_ids.mapped('price_include')):
+                    line.sale_line_ids.price_unit = -line.sale_line_ids.untaxed_amount_to_invoice
+            except UserError:
+                # a UserError here means the SO was locked, which prevents changing the taxes
+                # just ignore the error - this is a nice to have feature and should not be blocking
+                pass
```
Comment (smetl) [inline addons/sale/models/account_move.py:42]: You can't do that like this. Taxes are not so "easy". This whole downpayment feature is buggy regarding its taxes management so I think we should consider this as a limitation and let the user fixes the invoice manually. Someone in accounting is improving that.

Comment (guva-odoo) [reply addons/sale/models/account_move.py:42]: Ok, seems to be already fixed since I pushed anyway, thanks !

**Pull request closed without merging**